### PR TITLE
Fix the lack of support for the `blocking` option in the external source operator

### PR DESCRIPTION
**Pull request by JanuszL**
- restores support for `blocking` in the external source operator,
  that was removed when the batch size inference was introduced,
  making access to the input operator data earlier than in
  the operator's setup phase
- adjusts the error message in CachingList

## Category:
**Bug fix** (*non-breaking change which fixes an issue*)

## Description:
- restores support for `blocking` in the external source operator,
  that was removed when the batch size inference was introduced,
  making access to the input operator data earlier than in
  the operator's setup phase
- adjusts the error message in CachingList

### Affected modules and functionalities:
- input operator
- caching_list.h

### Key points relevant for the review:
- where the fix should be placed

### Tests:

- [ ] Existing tests apply
- [x] New tests added
  - test_external_source_dali.py:test_blocking
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by JanuszL** [comment]
!build

File: dali/test/python/test_external_source_dali.py [hunk 1]
```diff
@@ -741,3 +744,36 @@ def pipeline():
...
+
+    @pipeline_def
+    def test_pipeline():
+        jpegs = fn.external_source( dtype=types.UINT8, name="test_source", blocking=True)
+        decoded = fn.decoders.image(jpegs, device=device)
+        return decoded
+    root = os.path.join(DALI_EXTRA_PATH, 'db', 'single', 'jpeg')
+    file = glob.glob("**/*.jpg", root_dir=root)[0]
+    file = os.path.join(root, file)
+    data_to_feed = [np.fromfile(file, dtype=np.uint8) for _ in range(batch_size)]
```
Comment (mzient) [inline dali/test/python/test_external_source_dali.py:761]: I'd suggest using a much simpler (and faster) test - just feed a bunch of fixed numpy arrays, return them unchanged and compare.

Comment (JanuszL) [reply dali/test/python/test_external_source_dali.py]: Fixed

**Comment by JanuszL** [comment]
!build

File: dali/pipeline/operator/builtin/caching_list.h [hunk 2]
```diff
@@ -108,22 +108,26 @@ class CachingList {
   const T &PeekProphet() {
     if (prophet_ == full_data_.end())
       throw std::out_of_range(
-              "Attempted to peek element that doesn't exist. Add more elements to CachingList "
-              "before calling PeekProphet. Even the prophet can't see outside the event horizon.");
+              "Attempted to peek batch data that doesn't exist. Add more elements the DALI"
+              " input operator.");
```
Comment (szalpal) [inline dali/pipeline/operator/builtin/caching_list.h:112]:
```suggestion
              "Attempted to peek the data batch that doesn't exist. Add more elements to the DALI"
              " input operator.");
```
My only concern here is whether this error message would be always helpful. I mean, is it possible that after receiving this error message, adding more elements to the DALI input operator won't solve the error? If so, maybe this message should be rephrased to something like `Consider adding more elements [...]`?

File: dali/pipeline/operator/builtin/caching_list.h [hunk 3]
```diff
@@ -108,22 +108,26 @@ class CachingList {
...
 
 
   void AdvanceProphet() {
-    if (prophet_ == full_data_.end())
+    if (!CanProphetAdvance())
       throw std::out_of_range(
-              "Attempted to step over the last element in the list. This operation is forbidden. "
-              "Add more elements to CachingList before calling AdvanceProphet.");
+              "Attempted to move to the batch data that doesn't exist. Add more elements the DALI"
+              " input operator.");
```
Comment (szalpal) [inline dali/pipeline/operator/builtin/caching_list.h:121]:
```suggestion
              "Attempted to move to the data batch that doesn't exist. Add more elements to the DALI"
              " input operator.");
```

**Review by szalpal — APPROVED** [review]

File: dali/pipeline/operator/builtin/caching_list.h (hunk 3 above)
Comment (JanuszL) [reply dali/pipeline/operator/builtin/caching_list.h]: Fixed

File: dali/pipeline/operator/builtin/caching_list.h (hunk 2 above)
Comment (JanuszL) [reply dali/pipeline/operator/builtin/caching_list.h]: I'm afraid that now it talks about CachingList that is an implementation details, and prophet what make the user super confused. I'm open to suggestion how we can improve it further.

**Comment by JanuszL** [comment]
!build

**Pull request merged**